yktrace/src/tir.rs

```diff
@@ -19,50 +19,97 @@ use std::{collections::HashMap, convert::TryFrom, env, io::Cursor};
 #[cfg(debug_assertions)]
 use ykpack::BasicBlockIndex;
 pub use ykpack::Statement;
-use ykpack::{Body, Decoder, DefId, Local, Pack, SerU128, Terminator};
+use ykpack::{bodyflags, Body, Decoder, DefId, Local, Pack, SerU128, Terminator};
 
-// The SIR Map lets us look up a SIR body from the SIR DefId.
-// The map is unique to the executable binary being traced (i.e. shared for all threads).
 lazy_static! {
-    static ref SIR_MAP: HashMap<DefId, Body> = {
+    pub static ref SIR: Sir = {
         let ef = elf::File::open_path(env::current_exe().unwrap()).unwrap();
         let sec = ef.get_section(".yk_sir").expect("Can't find SIR section");
         let mut curs = Cursor::new(&sec.data);
         let mut dec = Decoder::from(&mut curs);
 
-        let mut sir_map = HashMap::new();
+        let mut bodies = HashMap::new();
+        let mut trace_head = None;
+        let mut trace_tails = Vec::new();
         while let Some(pack) = dec.next().unwrap() {
             match pack {
                 Pack::Body(body) => {
-                    let old = sir_map.insert(body.def_id.clone(), body);
+                    // Cache some locations that we need quick access to.
+                    if body.flags & bodyflags::TRACE_HEAD != 0 {
+                        assert_eq!(trace_head, None);
+                        trace_head = Some(body.def_id.clone());
+                    }
+
+                    if body.flags & bodyflags::TRACE_TAIL != 0 {
+                        trace_tails.push(body.def_id.clone());
+                    }
+
+                    let old = bodies.insert(body.def_id.clone(), body);
                     debug_assert!(old.is_none()); // should be no duplicates.
                 },
                 Pack::Debug(_) => (),
             }
         }
-        sir_map
+
+        assert!(trace_head.is_some());
+        assert_eq!(trace_tails.is_empty(), false);
+        let markers = SirMarkers { trace_head: trace_head.unwrap(), trace_tails };
+
+        Sir {bodies, markers}
     };
 }
 
+/// The serialised IR loaded in from disk. One of these structures is generated in the above
+/// `lazy_static` and is shared immutably for all threads.
+pub struct Sir {
+    /// Lets us map a DefId from a trace to a SIR body.
+    pub bodies: HashMap<DefId, Body>,
+    // Interesting locations that we need quick access to.
+    pub markers: SirMarkers
+}
+
+/// Contains the DefIds of interesting locations required for trace manipulation.
+pub struct SirMarkers {
+    /// The function which starts tracing and whose suffix gets trimmed off the top of traces.
+    /// There is only one such function: `yktrace::start_tracing()`.
+    ///
+    /// The compiler is made aware of this location by the `#[trace_head]` annotation.
+    pub trace_head: DefId,
+    /// Functions which stop tracing and whose prefix gets trimmed off the bottom of traces.
```

Comment:
Similarly, I assume the entirety of this function is removed from the end of the trace?